Fixing unchecked references to the JETPACK__PLUGIN_DIR constant

briancolinger opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes #

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* This should avoid fatal errors in cases when the JETPACK__PLUGIN_DIR constant is not defined.
* Check if the jetpack_protect_get_ip() function exists before using it.
* If the function does not exist, check if the JETPACK__PLUGIN_DIR constant is defined.
* If the constant is defined, require the shared-functions.php file, then call the jetpack_protect_get_ip() function.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Open wp shell for a Jetpack site. 
* `yarn docker:wp shell --url=DOMAIN`
* Verify the following wp shell commands return the proper data instead of throwing errors.
* return Automattic\Jetpack\Sync\Modules::get_module( 'options' );
* return Automattic\Jetpack\Sync\Functions::get_modules();

* Go to /projects/plugins/sync-tester
* `composer install`
* Login to your site's wp-admin.
* Check error-log for debug messages.


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* Sync: Removing references to the JETPACK__PLUGIN_DIR constant.

mdbitz commented on projects/packages/sync/src/class-listener.php:
```diff
@@ -417,8 +417,16 @@ public function get_actor( $current_filter, $args ) {
 		if ( $this->should_send_user_data_with_actor( $current_filter ) ) {
-			require_once JETPACK__PLUGIN_DIR . 'modules/protect/shared-functions.php';
-			$actor['ip']         = jetpack_protect_get_ip();
+			$ip = false;
```
As is this change will make it so that the ip of the actor is lost. If we don't want to move the shared-functions into a package at this time it would make sense to have a default value of:

$ip= isset( $_SERVER['REMOTE_ADDR'] ) ? $_SERVER['REMOTE_ADDR'] : '';

Which is what we do in the tracking package.

briancolinger replied:
Good point, updated the code.

mdbitz reviewed (changes requested):
@briancolinger what was your thougts on moving the shared-functions.php. I'm on the fence as I definitely believe the sync package is the wrong location for it. Also there is no other packages that it should live in at this time.

As long as we have a good default value we can move this forward.

kbrown9 commented:
The `shared-functions.php` file appears to include multiple functions that are specific to the Jetpack Protect module, so I don’t think there’s much value in moving all of those functions to a package.

I think the decision on whether to move the `jetpack_protect_get_ip()` function (and its dependencies) to a a package should be driven by whether we want Sync to always attempt to obtain the IP using the trusted header as it does now, or whether it’s acceptable for Sync to obtain the IP from the remote address value when Jetpack is not installed. (Good suggestion on using that as a default value @mdbitz.)

I’m in favor of maintaining the current behavior, so I think Sync should always attempt to obtain the ip from the trusted header, regardless of whether Jetpack is installed. So, I think we should investigate moving `jetpack_protect_get_ip()` to a package.

> I'm on the fence as I definitely believe the sync package is the wrong location for it. 

Generally, I’m not a fan of creating small utility packages that are used by only one other package, with no plans to use it in other packages. However in this case, I see that `jetpack_protect_get_ip()` is used a handful of times in the Jetpack plugin, and it does seem a bit illogical to call a function from the Sync package to get the ip.

What does everyone think of investigating creating a new IP package?

kbrown9 commented:
> I think the decision on whether to move the jetpack_protect_get_ip() function (and its dependencies) to a package should be driven by whether we want Sync to always attempt to obtain the IP using the trusted header as it does now, or whether it’s acceptable for Sync to obtain the IP from the remote address value when Jetpack is not installed.

The team discussed this, and we've decided that it's acceptable for Sync to obtain the IP from the remote address value when Jetpack is not installed. So, there's no need to move the `jetpack_protect_get_ip()` function to a new package.

mdbitz commented on projects/packages/sync/src/class-listener.php:
```diff
@@ -417,8 +417,16 @@ public function get_actor( $current_filter, $args ) {
 		if ( $this->should_send_user_data_with_actor( $current_filter ) ) {
-			require_once JETPACK__PLUGIN_DIR . 'modules/protect/shared-functions.php';
-			$actor['ip']         = jetpack_protect_get_ip();
+			$ip = isset( $_SERVER['REMOTE_ADDR'] ) ? $_SERVER['REMOTE_ADDR'] : '';
```
Would it make sense to simplify the complexity of the else { if () { statement into an else if { statement ?

} else if ( defined( 'JETPACK__PLUGIN_DIR' ) ) {

removes 2 addition lines and indenting.

mdbitz replied:
alternatively it could also be

if( function_exists( 'jetpack_protect_get_ip') || ( defined( 'JETPACK__PLUGIN_DIR' ) && include_once JETPACK__PLUGIN_DIR . 'modules/protect/shared-functions.php' ) ) {
    $ip = jetpack_protect_get_ip();
}


which would make it go from 8 lines to 3

briancolinger replied:
Great point! Updated the code.

briancolinger commented:
I've also included changes for the projects/packages/sync/src/class-functions.php and projects/packages/sync/src/modules/class-options.php files.

I see one other file that has a call to the JETPACK__PLUGIN_DIR constant: projects/packages/sync/vendor/automattic/jetpack-connection/legacy/class-jetpack-xmlrpc-server.php
@mdbitz does this file need to be updated as well?

kbrown9 commented:
> I see one other file that has a call to the JETPACK__PLUGIN_DIR constant: projects/packages/sync/vendor/automattic/jetpack-connection/legacy/class-jetpack-xmlrpc-server.php
> @mdbitz does this file need to be updated as well?

`JETPACK__PLUGIN_DIR` does need to be removed from that file. However, removing that is going to require a new API package, which is a significant change. Some details about a plan for the API package(s) are here: p9dueE-1H2-p2.

mdbitz commented:
@briancolinger as Kim outlined no steps at this time concerning JETPACK__PLUGIN_DIR usage in projects/packages/sync/vendor/automattic/jetpack-connection/legacy/class-jetpack-xmlrpc-server.php

Our focus is directly on the Sync package itself which would exclude vendor folder which is all of Sync's dependencies. Once we get the Sync package issues resolved we can help with the other packages that Kim has identified and is tracking in the project thread/task list.

mdbitz commented on projects/packages/sync/src/class-listener.php:
```diff
@@ -417,8 +417,11 @@ public function get_actor( $current_filter, $args ) {
 		if ( $this->should_send_user_data_with_actor( $current_filter ) ) {
-			require_once JETPACK__PLUGIN_DIR . 'modules/protect/shared-functions.php';
-			$actor['ip']         = jetpack_protect_get_ip();
+			if ( function_exists( 'jetpack_protect_get_ip' ) || ( defined( 'JETPACK__PLUGIN_DIR' ) && include_once JETPACK__PLUGIN_DIR . 'modules/protect/shared-functions.php' ) ) {
```
$ip= isset( $_SERVER['REMOTE_ADDR'] ) ? $_SERVER['REMOTE_ADDR'] : '';

It looks like we lost this line in the latest version to have the default value if jetpack_protect_get_ip is not available.

briancolinger replied:
Oopse! My mistake. Fixed.

zinigor replied:
I'm afraid that this line looks too clever. Can we make it more obvious please? I would carry the defined check up and call the `include` conditionally only if it's set, then we can check whether the function exists. 
Otherwise this `if` takes a while to understand.

mdbitz commented on projects/packages/sync/src/modules/class-options.php:
```diff
@@ -364,7 +364,10 @@ private function filter_theme_mods( &$value ) {
 		$url = get_site_icon_url();
 
-		require_once JETPACK__PLUGIN_DIR . 'modules/site-icon/site-icon-functions.php';
+		if ( defined( 'JETPACK__PLUGIN_DIR' ) ) {
```
This needs a little more work than a check on defined.  Looking at the site-icon-functions file it defines the function `jetpack_site_icon_url`

We need to rework the entire jetpack_sync_core_icon function so that we don't try calling `jetpack_site_icon_url` if we don't have the full jetpack installed.

briancolinger replied:
Fixed.

mdbitz commented on projects/packages/sync/src/class-functions.php:
```diff
@@ -22,7 +22,9 @@ class Functions {
 	 */
 	public static function get_modules() {
-		require_once JETPACK__PLUGIN_DIR . 'class.jetpack-admin.php';
+		if ( defined( 'JETPACK__PLUGIN_DIR' ) ) {
```
class.jetpack-admin.php defines the Jetapck_Admin class.

This change will lead to fatals as Jetpack_Admin will not be defined if the Sync package is being used without the full plugin.

I suggest that get_modules return an empty array in this case. Perhaps have the existing line 29 move up below the require statement and replace line 29 with return [];

briancolinger replied:
Fixed.

mdbitz reviewed (changes requested):
@briancolinger Similar to the work for ip we need to make sure that we aren't trying to access functionality that is no longer available if files aren't being required.

Left some notes for further action.

briancolinger commented:
I tried finding a way to test these changes and for the life of me, I can not figure out what I'm doing wrong.

I tried running this in wpsh on my wpcom sandbox:
`return Automattic\Jetpack\Sync\Modules::get_module( 'options' );`
It returns an array with modules listed as expected.

I tried copying over the changes manually. This may be where I'm going wrong.
I can't seem to get wpsh to recognize the changes. I may be doing this entirely wrong.

Maybe I was naively thinking I could easily test this by wrapping the instantiation inside a function that can be executed via wpsh. Maybe I haven't setup the proper amount of hoops to jump through yet?

mdbitz commented:
> I tried finding a way to test these changes and for the life of me, I can not figure out what I'm doing wrong.
> 
> I tried running this in wpsh on my wpcom sandbox:
> `return Automattic\Jetpack\Sync\Modules::get_module( 'options' );`
> It returns an array with modules listed as expected.
> 
> I tried copying over the changes manually. This may be where I'm going wrong.
> I can't seem to get wpsh to recognize the changes. I may be doing this entirely wrong.
> 
> Maybe I was naively thinking I could easily test this by wrapping the instantiation inside a function that can be executed via wpsh. Maybe I haven't setup the proper amount of hoops to jump through yet?

@briancolinger can you outline the tests you are trying to run. THe main thing here is the context we need to test behavior with and with out Jetpack.

1) On Jetpack test site run Automattic/Jetpack/Sync/Functions::get_modules(); from wp shell.
2) Verify the list of modules is returned.
3) on a non jetpack site. (install /include the sync package & dependencies) run Automattic/Jetpack/Sync/Functions::get_modules(); from wp shell.
4) verify that an empty array is returned.
5) Verify in both cases no errors, warnings, notices

Similar for the other updates. For non-jetpack sites we could possible use the connection demo project or we unit tests may be sufficient to verify the expected results.  E.g. Steps to manually tests that Jetpack is still working as expected. Then unit tests for the package by itself behavior.

briancolinger commented:
I'm going to circle back to this task tomorrow now that I have a better understanding of how to test these changes.

briancolinger commented:
> * on a non jetpack site. (install /include the sync package & dependencies) run Automattic/Jetpack/Sync/Functions::get_modules(); from wp shell.

@mdbitz I could use some help with this part of the testing instructions.

I was finally able to get the first part of testing on a Jetpack site figured out.
I was doing it wrong :)

The correct way is to have the Docker instance running, then load up wp shell... looking back... it's painfully obvious now.

briancolinger commented:
I think I'm still going to need help with the testing instructions on a non-jetpack site.

briancolinger commented:
I was able to get the sync-tester plugin files added and it even seems to build correctly!
Props to @kbrown9 for the sync-tester code.
I'll get the testing instructions updated next.

kbrown9 commented on projects/plugins/sync-tester/composer.json:
```diff
@@ -0,0 +1,26 @@
+{
```
I don't think we want to include this test plugin in the PR, because we don't want this merged into our codebase. It's only useful for testing the specific changes in this PR and isn't useful for testing anything else.

zinigor commented on projects/packages/sync/src/modules/class-options.php:
```diff
@@ -364,9 +364,13 @@ private function filter_theme_mods( &$value ) {
 
-		require_once JETPACK__PLUGIN_DIR . 'modules/site-icon/site-icon-functions.php';
+		$jetpack_url = \Jetpack_Options::get_option( 'site_icon_url' );
+		if ( function_exists( 'jetpack_site_icon_url' ) || ( defined( 'JETPACK__PLUGIN_DIR' ) && include_once JETPACK__PLUGIN_DIR . 'modules/site-icon/site-icon-functions.php' ) ) {
```
Oof, same here, this is too long and not very readable, can we fix it please?

briancolinger replied:
Agreed, this pattern was bothering me a bit too.
Fixed it.

zinigor reviewed (changes requested):
Thanks for working on these! I have left two nitpicks in addition to what Kim said about the plugin.

briancolinger commented:
My mistake about the sync-tester plugin. I've removed it.
I added it to the PR out of an abundance of caution... I didn't want to lose the code in case we needed later.

kbrown9 commented on projects/packages/sync/src/modules/class-options.php:
```diff
@@ -364,9 +364,14 @@ private function filter_theme_mods( &$value ) {
 
-		require_once JETPACK__PLUGIN_DIR . 'modules/site-icon/site-icon-functions.php';
+		$jetpack_url = \Jetpack_Options::get_option( 'site_icon_url' );
+		if ( function_exists( 'jetpack_site_icon_url' ) || defined( 'JETPACK__PLUGIN_DIR' ) ) {
```
@zinigor - what do you think of this approach? Is this what you were looking for in [your comment](https://github.com/Automattic/jetpack/pull/18608#discussion_r606308780)? I'm surprised to see the `include_once` called even when the function already exists. But, since it's an `include_once`, it should be fine.

mdbitz replied:
the reason for the include_once was to not have another if clause, to distinguish the include. No harm in having it if we want to be explicit.

if( defined( 'JETPACK__PLUGIN_DIR' ) ) {
    if( ! function_exists( 'jetpack_site_icon_url' ) ) {
        include_once JETPACK__PLUGIN_DIR . 'modules/site-icon/site-icon-functions.php';
    }
    $jetpack_url = jetpack_site_icon_url();
}

zinigor replied:
First of all, sorry for causing this fuss with the conditionals, but I really would like to see the code be more simple here. And I think Matthew's approach here aims to preserve the same logic that we had before we introduced the constant check.

So:

1. checking whether the constant is defined,
2. then checking whether the function is available - this check should be separate of the plugin check logic, because for some strange reason we could be in Jetpack context, but not have the file loaded;
3. then *requiring* the file as before, not including it (although here I don't know if it would matter that much - if the file is not there, the function call will still fatal after the `include`);

I hope that makes sense ^

zinigor replied:
Done in 24106d2.

zinigor commented:
Thanks for taking a look, y'all! I have made the change to the logic and merged `master` into the branch to keep it up to date.

kbrown9 reviewed (approved):
These changes look good to me. I tested with Jetpack, and everything seemed to work correctly. I also tested with Jetpack deactivated and the attached little test plugin which uses the Sync package, and no errors were generated.

[sync-tester.zip](https://github.com/Automattic/jetpack/files/6286683/sync-tester.zip)
